# Amazon Products Recommendations

Bu projede Amazon ürünlerine verilen müşteri puanları kullanılarak Popülariteye göre ürün tavsiye eden sistem geliştirilmektedir.
<br> Bu projede kullanılan veriler <a href="https://thecleverprogrammer.com/2021/03/23/amazon-recommendation-system-using-python/"
target="_blank">bu sayfadan</a> indirilmiştir.
<br><br>
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a9/Amazon_logo.svg/905px-Amazon_logo.svg.png" width="400px">

In [1]:
import pandas as pd
import numpy as np
import warnings 
warnings.filterwarnings("ignore")

In [2]:
df=pd.read_csv("ratings_Electronics (1).csv", header=None)
df.columns=['user_id', 'product_id','ratings','timestamp']

In [3]:
df.head()

user_id  product_id  ratings   timestamp
0   AKM1MP6P0OYPR  0132793040      5.0  1365811200
1  A2CX7LUOHB2NDG  0321732944      5.0  1341100800
2  A2NWSAGRHCP8N5  0439886341      1.0  1367193600
3  A2WNBOD3WNDNKT  0439886341      3.0  1374451200
4  A1GI0U4ZRJA8WN  0439886341      1.0  1334707200

In [4]:
# Kaç ürün var?
df["product_id"].nunique()

476002

In [5]:
# Ürünlere kaçar tane puan verilmiş?
df["product_id"].value_counts()

product_id
B0074BW614    18244
B00DR0PDNE    16454
B007WTAJTO    14172
B0019EHU8G    12285
B006GWO5WK    12226
              ...  
B004WL91KI        1
B004WL9FK4        1
B004WL9Q2Q        1
B004WL9R8O        1
BT008V9J9U        1
Name: count, Length: 476002, dtype: int64

In [6]:
# Ürünlere verilen puanların adet, toplam ve ortalama değerlerini hesaplayalım

product_grouped=df.groupby("product_id").agg({"ratings":[np.size, np.sum, np.mean]})

In [7]:
product_grouped

ratings                
              size   sum      mean
product_id                        
0132793040       1   5.0  5.000000
0321732944       1   5.0  5.000000
0439886341       3   5.0  1.666667
0511189877       6  27.0  4.500000
0528881469      27  77.0  2.851852
...            ...   ...       ...
BT008G3W52       1   5.0  5.000000
BT008SXQ4C       1   1.0  1.000000
BT008T2BGK       1   5.0  5.000000
BT008UKTMW      15  60.0  4.000000
BT008V9J9U       1   5.0  5.000000

[476002 rows x 3 columns]

In [8]:
# Ortalama puana göre sıralayalım

popular_products=product_grouped.sort_values(("ratings", "mean"), ascending=False)

In [9]:
popular_products.head()

ratings          
              size  sum mean
product_id                  
0132793040       1  5.0  5.0
B004SGH244       1  5.0  5.0
B004SQBRRM       1  5.0  5.0
B004SQC6O0       1  5.0  5.0
B004SQE6YI       1  5.0  5.0

In [10]:
# Toplam puanlara göre gruplayıp toplam puanı hesaplayalım

grouped_sum=product_grouped["ratings"]["sum"].sum()

In [11]:
#Ürünlere verilen toplam puanı, tüm ürünlere verilen toplam puana bölerek oranını hesaplayalım

popular_products["percentage"]=product_grouped["ratings"]["sum"].div(grouped_sum)*100

In [12]:
popular_products.head()

ratings           percentage
              size  sum mean           
product_id                             
0132793040       1  5.0  5.0   0.000016
B004SGH244       1  5.0  5.0   0.000016
B004SQBRRM       1  5.0  5.0   0.000016
B004SQC6O0       1  5.0  5.0   0.000016
B004SQE6YI       1  5.0  5.0   0.000016

In [13]:
# Orana göre sıralayıp en popüler ürünleri bulalım.

popular_products=popular_products.sort_values(("percentage"), ascending=False)
popular_products

ratings                    percentage
              size      sum      mean           
product_id                                      
B0074BW614   18244  81943.0  4.491504   0.261011
B00DR0PDNE   16454  64681.0  3.931020   0.206027
B007WTAJTO   14172  62697.0  4.424005   0.199707
B0019EHU8G   12285  58409.0  4.754497   0.186049
B006GWO5WK   12226  52751.0  4.314657   0.168026
...            ...      ...       ...        ...
B008UB07MC       1      1.0  1.000000   0.000003
B0008JFQ7K       1      1.0  1.000000   0.000003
B00008BX2O       1      1.0  1.000000   0.000003
B008UAQUAG       1      1.0  1.000000   0.000003
B000TFNDKK       1      1.0  1.000000   0.000003

[476002 rows x 4 columns]

In [14]:
# Popülerlik durumuna göre sıralama puanı verelim

popular_products["rank"]=popular_products["percentage"].rank(ascending=False)
popular_products

ratings                    percentage      rank
              size      sum      mean                     
product_id                                                
B0074BW614   18244  81943.0  4.491504   0.261011       1.0
B00DR0PDNE   16454  64681.0  3.931020   0.206027       2.0
B007WTAJTO   14172  62697.0  4.424005   0.199707       3.0
B0019EHU8G   12285  58409.0  4.754497   0.186049       4.0
B006GWO5WK   12226  52751.0  4.314657   0.168026       5.0
...            ...      ...       ...        ...       ...
B008UB07MC       1      1.0  1.000000   0.000003  463940.5
B0008JFQ7K       1      1.0  1.000000   0.000003  463940.5
B00008BX2O       1      1.0  1.000000   0.000003  463940.5
B008UAQUAG       1      1.0  1.000000   0.000003  463940.5
B000TFNDKK       1      1.0  1.000000   0.000003  463940.5

[476002 rows x 5 columns]

## Sonuç:
Yapılan Popülaritity analizine göre <b>B0074BW614</b> numaralı ürün toplam oyların %26.1'ini alarak en polüler ürün olarak belirlenmiştir. Ürün 18244 adet yoruma ortalama 4.49/5 puan almıştır.
<br>Ürürün kodu amazonda aratılınca artık satılmayan eski bir ürün olduğu görüldü. 
<br> Bu ürün: <b>Kindle Fire HD 7", Dolby Audio, Dual-Band Wi-Fi, 16 GB (Previous Generation - 2nd)</b>


In [15]:
popular_products.to_csv("popular_products.csv", index=None)